app/src/main/kotlin/at/bitfire/davdroid/sync/Syncer.kt

```diff
@@ -62,62 +66,119 @@ abstract class Syncer {
     lateinit var context: Context
 
     @Inject
-    lateinit var db: AppDatabase
+    lateinit var collectionRepository: DavCollectionRepository
 
     @Inject
     lateinit var logger: Logger
 
+    @Inject
+    lateinit var serviceRepository: DavServiceRepository
+
+    abstract val authority: String
+    abstract val serviceType: String
+
+    val accountSettings by lazy { accountSettingsFactory.forAccount(account) }
+    val httpClient = lazy { HttpClient.Builder(context, accountSettings).build() }
+
+    private val dbCollections = mutableMapOf<HttpUrl, Collection>()
 
     /**
-     * Creates and/or deletes local collections (calendars, address books, etc) and updates them
-     * with remote information. Then syncs the actual entries (events, tasks, contacts, etc) of all
-     * collections.
+     * Creates, updates and/or deletes local resources (calendars, address books, etc) according to
+     * remote collection information. Then syncs the actual entries (events, tasks, contacts, etc)
+     * of the remaining up-to-date resources.
      */
-    abstract fun sync(account: Account, extras: Array<String>, authority: String, httpClient: Lazy<HttpClient>, provider: ContentProviderClient, syncResult: SyncResult)
-
-    fun onPerformSync(
-        account: Account,
-        extras: Array<String>,
-        authority: String,
-        syncResult: SyncResult
-    ) {
-        logger.log(Level.INFO, "$authority sync of $account initiated", extras.joinToString(", "))
+    fun sync(provider: ContentProviderClient) {
+        // 0. resource specific preparations
+        if (!prepare(provider)) {
+            logger.log(Level.WARNING, "Failed to prepare sync. Won't run sync.")
+            return
+        }
+
+        // 1. find resource collections to be synced
+        serviceRepository.getByAccountAndType(account.name, serviceType)?.let { service ->
+            for (dbCollection in getSyncCollections(service.id))
+                dbCollections[dbCollection.url] = dbCollection
+        }
+
+        // 2. update/delete local resources and determine new (unknown) remote collections
+        val newDbCollections = HashMap(dbCollections)   // create a copy
+        for (localCollection in localCollections(provider)) {
```

Comment:
Currently, we don't use "non-sync" collections at all. So for instance a calendar that is not selected for synchronization is currently not created as local calendar with SYNC=0, but not even created.

See for instance the old `CalendarSyncer`, which fills its `remoteCollections` like that:

```
if (service != null)
    for (collection in db.collectionDao().getSyncCalendars(service.id))   // <- getSyncCalendars
        remoteCollections[collection.url] = collection
```

I'd like to leave it like that. So I guess we don't need `localCollections` at all, and only `localSyncCollections`?

Reply:
Yes, I think that should be possible. I was wondering wether we could do this, but could not oversee the implications changing it would have.